single_kernel_mongo/managers/mongo.py

```diff
@@ -0,0 +1,415 @@
+
+from single_kernel_mongo.config.literals import Substrates
+from single_kernel_mongo.core.structured_config import MongoDBRoles
+from single_kernel_mongo.exceptions import DatabaseRequestedHasNotRunYetError, SetPasswordError
+from single_kernel_mongo.lib.charms.data_platform_libs.v0.data_interfaces import (
+    DatabaseProviderData,
+)
+from single_kernel_mongo.managers.k8s import K8sManager
+from single_kernel_mongo.state.charm_state import CharmState
+from single_kernel_mongo.utils.mongo_config import (
+    EMPTY_CONFIGURATION,
+    MongoConfiguration,
+)
+from single_kernel_mongo.utils.mongo_connection import MongoConnection, NotReadyError
+from single_kernel_mongo.utils.mongodb_users import (
+    OPERATOR_ROLE,
+    BackupUser,
+    MongoDBUser,
+    MonitorUser,
+    OperatorUser,
+)
+from single_kernel_mongo.workload.mongodb_workload import MongoDBWorkload
+
+if TYPE_CHECKING:
+    from single_kernel_mongo.core.operator import OperatorProtocol
+
+logger = logging.getLogger(__name__)
+
+
+class MongoManager(Object):
+    """Manager for Mongo related operations."""
+
+    def __init__(
+        self,
+        dependent: OperatorProtocol,
+        workload: MongoDBWorkload,
+        state: CharmState,
+        substrate: Substrates,
+    ) -> None:
+        super().__init__(parent=dependent, key="managers")
+        self.charm = dependent.charm
+        self.workload = workload
+        self.state = state
+        self.substrate = substrate
+        pod_name = self.model.unit.name.replace("/", "-")
+        self.k8s = K8sManager(pod_name, self.model.name)
+
+    def mongod_ready(self, uri: str | None = None) -> bool:
+        """Is MongoDB ready and running?"""
+        config = EMPTY_CONFIGURATION
+        actual_uri = uri or "localhost"
+        with MongoConnection(config, actual_uri, direct=True) as direct_mongo:
+            return direct_mongo.is_ready
+
+    def set_user_password(self, user: MongoDBUser, password: str) -> str:
+        """Sets the password for a given username and return the secret id.
+
+        Raises:
+            SetPasswordError
+        """
+        with MongoConnection(self.state.mongo_config) as mongo:
+            try:
+                mongo.set_user_password(user.username, password)
+            except NotReadyError:
+                raise SetPasswordError(
+                    "Failed changing the password: Not all members healthy or finished initial sync."
+                )
+            except PyMongoError as e:
+                raise SetPasswordError(f"Failed changing the password: {e}")
+
+        return self.state.set_user_password(user, password)
+
+    def initialise_replica_set(self) -> None:
+        """Initialises the replica set."""
+        with MongoConnection(self.state.mongo_config, "localhost", direct=True) as direct_mongo:
+            direct_mongo.init_replset()
+            self.state.app_peer_data.replica_set_hosts = [self.state.unit_peer_data.host]
```

Comment:
might we want to handle these errors?

            if e.code not in (13, 23):  # Unauthorized, AlreadyInitialized
                # Unauthorized error can be raised only if initial user were
                #     created the step after this.
                # AlreadyInitialized error can be raised only if this step
                #     finished.
                logger.error("Cannot initialize replica set. error=%r", e)
                raise e